### PR TITLE
Mark functions either noreturn or returns_nonnull

### PR DESCRIPTION
**Pull request by nickdesaulniers**
to silence 4 null pointer warnings from static
analysis #107

Also, `_Noreturn` [seems to be](http://stackoverflow.com/a/21646860/1027966) part of C11, and [MSVS has](http://msdn.microsoft.com/en-us/library/k6ktzx3s.aspx) `__declspec(noreturn)`. Not sure we want C11 features or if the M$ thing id C++ only?  Closing #108 in favor of this.

### CODE REVIEW

File: include/h2o/memory.h [hunk 1]
```diff
@@ -34,6 +34,18 @@ extern "C" {
 
 #define H2O_STRUCT_FROM_MEMBER(s, m, p) ((s *)((char *)(p)-offsetof(s, m)))
 
+#if defined(__clang__) || defined(__GNUC__) && (__GNUC__ > 2 || __GNUC__ == 2 && __GNUC_MINOR__ >= 5)
```
Comment (nickdesaulniers) [inline include/h2o/memory.h]: Also, this line is super long, wasn't sure if it was ok to split up preprocessor statements across multiple lines.

**Comment by kazuho** [comment]
Thank you for the PR.

It looks basically fine.

> Also, _Noreturn seems to be part of C11, and MSVS has __declspec(noreturn).

Good catch!  It would be preferable to use _Noreturn in case of C11 (you can use `__STDC_VERSION__` to check if the compiler supports the feature ([ref](http://en.wikipedia.org/wiki/C11_%28C_standard_revision%29#Version_detection)).

Setting flags for MSVC is also favorable but not mandatory at this point.

I have left in-line comments for other issues.

**Comment by nickdesaulniers** [comment]
> please check the GCC version before using returns_nonnull as it seems to have been introduced pretty recently (ref).

The patch was written in Oct 8 2013.  Depending on how long it took to merge, it looks like that would be GCC 4.8.2 and greater.  I was hoping there was a better way to conditionally check for support than guessing and checking the compiler version. :-/

**Comment by nickdesaulniers** [comment]
I have incorporated the feedback.  Maybe you would like to have a newer version of GCC check for `returns_nonnull`?

**Comment by nickdesaulniers** [comment]
Hows that look?

**Comment by kazuho** [comment]
Thanks a lot!  merged to master.

**Pull request merged**